enhancements/authentication/service-ca-cert-generation-for-statefulset-pods.md

```diff
@@ -0,0 +1,344 @@
+
+  (Reportedly some OSes do actually query `localhost` on DNS on the net:
+  https://archive.cabforum.org/pipermail/public/2015-June/005673.html )
+
+  If a client wants to connect to _only_ a “true localhost” service,
+  a service CA-signed certificate does not provide that property.
+  The best way to do that is to do it directly inside the process,
+  without using TLS (and the associated cryptography overhead) at all.
+  Second best is to rely on known properties of the environment,
+  i.e. to assume that 127.0.0.1 is localhost-only
+  and to use a non-TLS connection on that address.
+
+  If a _pod_ had to use TLS for localhost for some reason, the closest to a right way to do that
+  (relying on TLS to verify that only loopback connections happen) is to:
+  - Create a temporary CA (specific to the individual pod for)
+  - Generate, and have that CA sign a localhost certificate (specific to the individual pod)
+  - Set up the CA certificate as a trusted root
+  - _Irreversibly erase_ the CA private key
+  - Set up the pod with the generated localhost certificate+private key.
+  - (This doesn’t benefit from service-ca involvement at all.)
+  
+  Similarly single-purpose CA could be set up for NodeIP services.
+
+- Generate a single certificate that signs multiple ~unrelated services as subjects
+  (combining the StatefulSet per-pod subjects and ClusterIP services,
+  or a ClusterIP service with NodePort service):
+
+  In the most general case, this is not possible at all, because services choose pods by pod labels,
+  i.e. the set of services served by a single pod at once (which require a shared certificate)
+  can only be determined once such a pod is created —
+  and at that point it’s too late to generate a certificate for that pod.
+
+  If the different services are served on different ports,
+  they can trivially use different certificates.
+  Even for a single port exposed via the different services,
+  the server should be able to use different certificates based on SNI nowadays
+  (e.g. via https://golang.org/pkg/crypto/tls/#Config.Certificates ).
+
+## Proposal
+
+Similarly to the existing annotation on services,
+a `service.beta.openshift.io/serving-cert-secret-name: $secretName` annotation on a `StatefulSet`
+requests the `service-ca` controller to generate a single secret `$secretName`
+with a set of individual certificate+key pairs named `{0,1,2,3,…}.{crt,key}`.
+The subject of `$podID.crt` is the pod’s unique DNS identity,
+`${statefulset.name}-${podID}.${service.name}.${service.namespace}.svc` .
+Each pod from a `StatefulSet` can then pick and use its own certificate.
+
+Creating a `StatefulSet` triggers both a `Secret` creation and creating the `StatefulSet` pods,
+which race against each other.
+So, each individual pod **must** fail if a certificate for its individual identity is not found.
+We also generate a few more certificate+key pairs than necessary,
+to allow scaling up without such pod failures.
```

Comment:
how many?

Reply:
What happens if we scale too far?

Reply:
nvmd about second question, it's answered below

Reply:
> The implementation currently hard-codes creating 5 extra certificates, or 30% more than spec.replicas, whichever is larger.

This is just a guess for how quickly already-operational StatefulSets are likely to scale up, with an assumption that generating new key pairs is cheap enough not to worry about; if that could become a bottleneck (e.g. if a FIPS-mode random number source were noticeably slow), this might require extra tuning or more work to avoid unnecessarily re-generating valid certificates in large batches.